File: GO/week2/ex04-calc/calc_test.go
package main

import "testing"

func TestCalculateFullCoverage(t *testing.T) {
	// Defining the test cases (8 paths)
	tests := []struct {
		name       string
		expression string
		want       float64
		wantErr    bool
	}{
		{"Addition", "10 + 5", 15, false},
		{"Subtraction", "20 - 8", 12, false},
		{"Multiplication", "4 * 3", 12, false},
		{"Division", "10 / 2", 5, false},
		{"Division by Zero", "10 / 0", 0, true},
		{"Invalid Format", "5 +", 0, true},
		{"Unknown Operator", "5 % 2", 0, true},
		{"Non-numeric Input", "abc + 2", 0, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := Calculate(tt.expression)

			// Checking if we expected an error and if we actually got one
			if (err != nil) != tt.wantErr {
				t.Errorf("Calculate() error = %v, wantErr %v", err, tt.wantErr)
				return
			}

			// Checking if the numeric result is correct
			if got != tt.want {
				t.Errorf("Calculate() got = %v, want %v", got, tt.want)
			}
		})
	}
}

File: javascript/week2/ex04-calc/calc.test.js
// @ts-nocheck
const { calculate } = require('./calc');

describe('Calculator - Full Test Coverage', () => {
  // Happy Paths
  test('should handle addition: "10 + 5" -> 15', () => {
    expect(calculate("10 + 5")).toBe(15);
  });

  test('should handle subtraction: "20 - 8" -> 12', () => {
    expect(calculate("20 - 8")).toBe(12);
  });

  test('should handle multiplication: "4 * 3" -> 12', () => {
    expect(calculate("4 * 3")).toBe(12);
  });

  test('should handle division: "10 / 2" -> 5', () => {
    expect(calculate("10 / 2")).toBe(5);
  });

  // Sad Paths (Error Handling)
  test('should error on division by zero', () => {
    expect(() => calculate("10 / 0")).toThrow("Division by zero not allowed");
  });

  test('should error on invalid format (missing parts)', () => {
    expect(() => calculate("5 +")).toThrow("Invalid format");
  });

  test('should error on unknown operator', () => {
    expect(() => calculate("5 % 2")).toThrow("Unknown operator");
  });

  test('should error on non-numeric inputs', () => {
    expect(() => calculate("abc + 2")).toThrow("Invalid numbers");
  });
});